### ISSUE TITLE
[Bug Report] Improving the gallery view performance for large collections

### ISSUE THREAD

**Issue by esanchezm** [stashapp/stash#2076]
I have a big collection, including more than 7M images in 14k galleries. I'm having issues with the gallery view as it takes a long time to fetch the images belonging to a gallery. The operation takes around 15 seconds so I get a timeout and nothing shows.

I tried to debug the SQL query and it's basically performing a full scan on the `images` table. That's obviously fine in most of the cases but with a 7M rows table it isn't.

```
explain query plan SELECT DISTINCT images.id FROM images LEFT JOIN galleries_images AS galleries_join ON galleries_join.image_id = images.id LEFT JOIN galleries ON galleries_join.gallery_id = galleries.id WHERE (galleries.id IN (14664)) GROUP BY images.id  HAVING (count(distinct galleries.id) IS 1) ORDER BY images.path ASC LIMIT 40 OFFSET 0;
QUERY PLAN
|--SCAN TABLE images
|--SEARCH TABLE galleries_images AS galleries_join USING INDEX index_galleries_images_on_image_id (image_id=?)
|--SEARCH TABLE galleries USING INTEGER PRIMARY KEY (rowid=?)
|--USE TEMP B-TREE FOR DISTINCT
`--USE TEMP B-TREE FOR ORDER BY
```

So, I think this is something that could be fixed by using `INNER JOIN` instead of `LEFT JOIN`:

```
sqlite> explain query plan SELECT DISTINCT images.id FROM images INNER JOIN galleries_images AS galleries_join ON galleries_join.image_id = images.id INNER JOIN galleries ON galleries_join.gallery_id = galleries.id WHERE (galleries.id IN (14664)) GROUP BY images.id  HAVING (count(distinct galleries.id) IS 1) ORDER BY images.path ASC LIMIT 40 OFFSET 0;
QUERY PLAN
|--SEARCH TABLE galleries USING INTEGER PRIMARY KEY (rowid=?)
|--SEARCH TABLE galleries_images AS galleries_join USING INDEX index_galleries_images_on_gallery_id (gallery_id=?)
|--SEARCH TABLE images USING INTEGER PRIMARY KEY (rowid=?)
|--USE TEMP B-TREE FOR GROUP BY
|--USE TEMP B-TREE FOR DISTINCT
`--USE TEMP B-TREE FOR ORDER BY
```

I suppose there are good reasons behind using a `LEFT JOIN`, but I think this should be fixed so Stash is usable for large/huge collections.

If I'm not mistaken, this is the offending line: https://github.com/stashapp/stash/blob/develop/pkg/sqlite/image.go#L17

**Comment by InfiniteTF** [comment]
Your analysis sounds correct. I can't think of any reason to use `LEFT JOIN` given that we're querying for images attached to a gallery. Any images without a gallery are by definition not a match.

**Comment by esanchezm** [comment]
Thanks @InfiniteTF :smile: 

I tried changing the line I referenced but apparently that's not enough. It looks like it requires a bigger refactor :disappointed:

**Comment by WithoutPants** [comment]
Resolved by #2083
